pkg/controllers/nodeclaim/tagging/controller.go

```diff
@@ -0,0 +1,179 @@
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Extract resource ID from node spec
+	resourceID := path.Base(nodeClaim.Status.ProviderID)
+	if resourceID == "." || resourceID == "/" {
+		return reconcile.Result{}, fmt.Errorf("failed to extract resource ID from providerID")
+	}
+
+	// Check if instance has already been tagged
+	if isTagged, err := c.isInstanceTagged(ctx, nodeClaim, resourceID); err != nil || isTagged {
+		return reconcile.Result{}, err
+	}
+
+	// Tag instance
+	err := c.tagInstance(ctx, nodeClaim, resourceID)
+	if err != nil {
+		return reconcile.Result{Requeue: true}, client.IgnoreNotFound(err)
+	}
+
+	logging.FromContext(ctx).
+		With("node", req.NamespacedName.Name).
+		With("instanceID", resourceID).
+		Debugf("updated ec2 instance Name tag")
+
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Builder(_ context.Context, m manager.Manager) corecontroller.Builder {
+	return corecontroller.Adapt(
+		controllerruntime.
+			NewControllerManagedBy(m).
+			For(&corev1beta1.NodeClaim{}).
+			WithEventFilter(onlyTaggableInstances()),
+	)
+}
+
+func onlyTaggableInstances() predicate.Predicate {
+	return predicate.NewPredicateFuncs(func(object client.Object) bool {
+		nodeClaim := object.(*corev1beta1.NodeClaim)
+
+		// Instance has already been tagged
+		if val, ok := nodeClaim.Annotations[v1beta1.AnnotationInstanceTagged]; ok && val == "true" {
+			return false
+		}
+
+		// Node name is not yet known
+		if nodeClaim.Status.NodeName == "" {
+			return false
+		}
+
+		return true
+	})
+}
+
+func (c *Controller) isInstanceTagged(ctx context.Context, nodeClaim *corev1beta1.NodeClaim, resourceID string) (bool, error) {
+	tag, err := c.getTagValue(resourceID, "Name")
+	if err != nil {
+		return false, err
+	}
+
+	// Check if the instance has the default tag (karpenter.sh/provisioner-name/*). If so it has either:
+	//   a) Already been tagged by us but it is not reflected on the API server
+	//   b) Tagged by some other process. In this case we will respect that tag.
+	if !c.defaultTagRgx.MatchString(tag) {
+		nodeClaim.Annotations[v1beta1.AnnotationInstanceTagged] = "true"
+		err := c.kubeClient.Update(ctx, nodeClaim)
+		return true, client.IgnoreNotFound(err)
+	}
+
+	return false, nil
+}
+
+func (c *Controller) getTagValue(resourceID string, tagKey string) (string, error) {
+	var tag *ec2.TagDescription
+	err := c.ec2.DescribeTagsPages(&ec2.DescribeTagsInput{
+		Filters: []*ec2.Filter{{
+			Name:   aws.String("resource-id"),
+			Values: []*string{&resourceID},
+		}},
+	}, func(output *ec2.DescribeTagsOutput, _ bool) bool {
+		for _, t := range output.Tags {
+			if *t.Key == tagKey {
+				tag = t
+				return false
+			}
+		}
+		return true
+	})
+	if err != nil {
+		return "", err
+	}
+
+	return lo.Ternary(tag == nil, "", *tag.Value), nil
+}
+
+func (c *Controller) tagInstance(ctx context.Context, nodeClaim *corev1beta1.NodeClaim, resourceID string) error {
```

Comment:
I think the intention of tagging the instance is that we tag the instance with the node name instead of the nodeClaim name. These values will be different. I think, ideally, we can add both to the instance so that we get the following tags on the instance

```yaml
tags:
  karpenter.sh/nodeclaim: nodeClaim.Status.Name
  Name: nodeClaim.Status.NodeName
```

Reply:
Right now I think it is being tagged with the node name, the entire `nodeClaim` is passed in to also update it's annotation but like you suggested in a different comment I'm going to refactor that out into the reconcile loop. Would we be able to annotate with the node claim name as part of the create fleet call rather than as part of this controller?